handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java

```diff
@@ -0,0 +1,187 @@
+    }
+
+    static void configureAlpn(JdkSslEngine engine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+        List<String> supportedProtocols = applicationNegotiator.protocols();
+
+        SSLEngine wrapped = engine.getWrappedEngine();
+        SSLParameters params = wrapped.getSSLParameters();
+        setApplicationProtocols(params, supportedProtocols);
+        wrapped.setSSLParameters(params);
+
+        if (!wrapped.getUseClientMode()) {
+            installSelector(engine, applicationNegotiator);
+        }
+    }
+
+    static String getApplicationProtocol(SSLEngine sslEngine) {
+        try {
+            return (String) GET_APPLICATION_PROTOCOL.invoke(sslEngine);
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void setApplicationProtocols(SSLParameters parameters, List<String> supportedProtocols) {
+        String[] protocolArray = supportedProtocols.toArray(EmptyArrays.EMPTY_STRINGS);
+        try {
+            SET_APPLICATION_PROTOCOL.invoke(parameters, new Object[]{ protocolArray });
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void installSelector(final JdkSslEngine sslEngine,
+                                        final JdkApplicationProtocolNegotiator applicationProtocolNegotiator) {
+        checkNotNull(applicationProtocolNegotiator, "applicationProtocolNegotiator");
+
+        Object biFunction = AccessController.doPrivileged(new PrivilegedAction<Object>() {
+            @Override
+            public Object run() {
+                return Proxy.newProxyInstance(Java9SslUtils.class.getClassLoader(), new Class[]{ BI_FUNCTION_CLASS },
+                        new AlpnSelector(sslEngine, applicationProtocolNegotiator));
+            }
+        });
+
+        try {
+            SET_HANDSHAKE_APPLICATION_PROTOCOL_SELECTOR.invoke(sslEngine.getWrappedEngine(), biFunction);
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static final class AlpnSelector implements InvocationHandler {
+        private final SSLEngine jdkSslEngine;
+        private final JdkApplicationProtocolNegotiator applicationNegotiator;
+        private final Set<String> protocols;
+
+        AlpnSelector(JdkSslEngine jdkSslEngine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+            this.jdkSslEngine = jdkSslEngine;
+            this.applicationNegotiator = applicationNegotiator;
+            protocols = Collections.unmodifiableSet(new LinkedHashSet<String>(applicationNegotiator.protocols()));
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            // Assert that it is actually called via the BiConsumer that we expect.
+            assert "apply".equals(method.getName());
```

Comment:
Include what the name was?  i.e. 

`assert "apply".equals(method.getName()); : "unexpected method " + method.getName();`